kyo-core/shared/src/main/scala/kyo/internal/LayerMacros.scala

```diff
@@ -0,0 +1,238 @@
+                    Debug: $debugFold
+                    """
+                        )
+                    }
+                )
+
+                exprFold.asInstanceOf[Expr[Layer[Target, ?]]]
+
+            case _ => report.errorAndAbort("No layers found. Did you forget to provide them?".red)
+        end match
+    end makeImpl
+
+    def layerToNode(using Quotes)(expr: Expr[Layer[?, ?]]): Node[reflect.TypeRepr, Expr[Layer[?, ?]]] =
+        import reflect.*
+        expr match
+            case '{ $layer: Layer[out, s] } =>
+                val inputs  = extractInputsFromPending(TypeRepr.of[s])
+                val outputs = flattenAnd(TypeRepr.of[out])
+                Node(inputs, outputs, expr)
+        end match
+    end layerToNode
+
+    def extractInputsFromPending(using Quotes)(typeRepr: reflect.TypeRepr): Set[reflect.TypeRepr] =
+        flattenAnd(typeRepr).flatMap(extractEnvs)
+
+    object AsType:
+        def unapply(using Quotes)(typeRepr: reflect.TypeRepr): Option[Type[?]] =
+            Some(typeRepr.asType)
+
+    def extractEnvs(using Quotes)(typeRepr: reflect.TypeRepr): Set[reflect.TypeRepr] =
+        import reflect.*
+        typeRepr.asType match
+            case '[Envs[tpe]] => flattenAnd(TypeRepr.of[tpe])
+            case _            => Set.empty
+    end extractEnvs
+
+    def flattenAnd(using Quotes)(typeRepr: reflect.TypeRepr): Set[reflect.TypeRepr] =
+        import reflect.*
+        typeRepr match
+            case AndType(left, right) =>
+                flattenAnd(left) ++ flattenAnd(right)
+            case typeRepr =>
+                Set(typeRepr)
+        end match
+    end flattenAnd
+end LayerMacros
+
+enum LayerLike[+A]:
+    case And(left: LayerLike[A], right: LayerLike[A])
+    case To(left: LayerLike[A], right: LayerLike[A])
+    case Value(value: A)
+    case Empty
+
+    infix def and[A1 >: A](that: LayerLike[A1]): LayerLike[A1] = And(this, that)
+    infix def to[A1 >: A](that: LayerLike[A1]): LayerLike[A1]  = To(this, that)
+
+    def fold[B](andCase: (B, B) => B, toCase: (B, B) => B, valueCase: A => B, emptyCase: => B): B =
+        this match
+            case And(left, right) =>
+                andCase(left.fold(andCase, toCase, valueCase, emptyCase), right.fold(andCase, toCase, valueCase, emptyCase))
+            case To(left, right) =>
+                toCase(left.fold(andCase, toCase, valueCase, emptyCase), right.fold(andCase, toCase, valueCase, emptyCase))
+            case Value(value) => valueCase(value)
+            case Empty        => emptyCase
+end LayerLike
+
+object LayerLike:
+    def apply[A](value: A): LayerLike[A]            = Value(value)
+    given [A]: CanEqual[LayerLike[A], LayerLike[A]] = CanEqual.derived
+
+final case class Node[Key, Value](inputs: Set[Key], outputs: Set[Key], value: Value)
+object Node:
+    given [Key, Value]: CanEqual[Node[Key, Value], Node[Key, Value]] = CanEqual.derived
+
+enum Validated[+E, +A]:
```

Comment:
I should have a proper public impl for this but I think it's ok to keep it internal for now!